### PR TITLE
feat: Helm chart

### PR DESCRIPTION
**Pull request by Yann-J**
## Description

This is a proper Helm chart to run it on Kubernetes, as a more flexible alternative to the existing deployment definitions and setup scripts. It uses the bitnami mysql chart as a dependency to install MySQL, so all these values can be overridden.

To install, just run:

```sh
# Install mysql chart dependency
helm dependency update

# Install chart
helm upgrade --install --set mysql.auth.rootPassword=xxxx my-fineract helm/fineract
```

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [x] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/api-docs/apiLive.htm with details of any API changes

- [x] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

**Comment by edcable** [comment]
@Yann-J It was nice speaking to you the other day. @vidakovic our release manager would love to include this in the forthcoming Fineract 1.5 release - are you able to address the failing build?

File: helm/fineract/values.yaml
Comment (vorburger) [inline helm/fineract/values.yaml]: These default password here scare me, a lot... people will NOT change these - they never do.

There must a better way to do this, with Helm; what would be the Helm equivalent of what we're doing in https://github.com/apache/fineract/blob/develop/kubernetes/kubectl-startup.sh#L22 with `kubectl create secret generic fineract-tenants-db-secret --from-literal=username=root --from-literal=password=$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c 16)` ?

**Review by vorburger — CHANGES_REQUESTED** [review]
No Go from me for hard-coded default passwords - there must be a better solution?

More minor: How about move this into `kubernetes/helm/` instead of (root) `helm/`?

Lastly, could you include the how to use doc in the README instead of PR desc?

**Comment by vorburger** [comment]
@BLasan have you noticed this PR? It looks like something you could be interested in helping to review... 😄 

I was thinking about something when reviewing #1653: That (resource limits) should be done here as well? More importantly, we're now maintaining Kubernetes YAML in 2 places (original raw Kube YAML, and Helm chart templates). That's perhaps not ideal... you guys should figure out how you want to do this going forward? Keep both? Replace the original YAML with the Helm templates, and subsequently remove the originals? Up to you - work together.

**Comment by BLasan** [comment]
> @BLasan have you noticed this PR? It looks like something you could be interested in helping to review... 
> 
> I was thinking about something when reviewing #1653: That (resource limits) should be done here as well? More importantly, we're now maintaining Kubernetes YAML in 2 places (original raw Kube YAML, and Helm chart templates). That's perhaps not ideal... you guys should figure out how you want to do this going forward? Keep both? Replace the original YAML with the Helm templates, and subsequently remove the originals? Up to you - work together.

It would be great if we could add resource limits in order to run the image using minimum number of resources in the cluster. (Cluster resources will not be over eaten by then) If someone wants to run the product in k8s without cloning it, they could use the helm charts directly (after releasing this) What if we could move these helm implementations to a new repository and maintain it there? Don't know whether this is a good idea. But we could maintain this and can do releases separately.

File: helm/fineract/templates/fineract-server-deployment.yml
Comment (BLasan) [inline helm/fineract/templates/fineract-server-deployment.yml:79]: You can use separate paths to check liveness and readiness probes. We have added configurations in to spring boot. See https://github.com/apache/fineract/pull/1649

File: helm/fineract/templates/fineract-server-deployment.yml
Comment (BLasan) [inline helm/fineract/templates/fineract-server-deployment.yml:75]: I think you have missed this env.
```
        - name: FINERACT_DEFAULT_TENANTDB_PWD
          valueFrom:
            secretKeyRef:
              name: fineract-tenants-db-secret
              key: password
```

File: helm/fineract/values.yaml
Comment (BLasan) [reply helm/fineract/values.yaml]: I don't know whether the following line would generate a proper random password. You can try this to generate the mysql password @Yann-J 

`password: {{ randAlphaNum 64 | quote }}`

Or else we can point a particular secret name to the get password. But in order to do that we need to create that secret first before installing helm chart
Refer: https://stackoverflow.com/questions/56170052/how-not-to-overwrite-randomly-generated-secrets-in-helm-templates

**Review by BLasan — CHANGES_REQUESTED** [review]

File: helm/fineract/values.yaml
Comment (Yann-J) [reply helm/fineract/values.yaml]: Hello there, sorry for my delay here!

We can't use `randAlphaNum` inside the chart because that would cause the value to be re-generated each time we deploy. However there's no problem to keep that value empty and implement a runtime warning to instruct users to provide a value.

I'm not so sure that "people will NOT change these" is really true any more with Helm since it's just so easy to do... but for extra protection it makes sense to force it, I can do that.

Comment (BLasan) [reply helm/fineract/values.yaml]: Agree with that. I think there was the same issue in mysql deployment in local cluster. I've fixed that. This issue was raised in a previous PR https://github.com/apache/fineract/pull/1583

File: helm/fineract/templates/fineract-server-deployment.yml
Comment (Yann-J) [reply helm/fineract/templates/fineract-server-deployment.yml:75]: I don't think so, this one is handled with:

```
          envFrom:
            - secretRef:
                name: {{ .Release.Name | quote }}
```

File: helm/fineract/templates/fineract-secret.yaml
Comment (BLasan) [inline helm/fineract/templates/fineract-secret.yaml:22]: Use `{{ .Release.Name }}-secret`.

File: helm/fineract/templates/fineract-server-deployment.yml
Comment (BLasan) [inline helm/fineract/templates/fineract-server-deployment.yml:59]: Use the secret name (this is bit confusing)

**Review by BLasan — CHANGES_REQUESTED** [review]

File: helm/fineract/templates/fineract-server-deployment.yml
Comment (Yann-J) [reply helm/fineract/templates/fineract-server-deployment.yml:79]: 👍 done!

File: helm/fineract/templates/fineract-secret.yaml
Comment (Yann-J) [reply helm/fineract/templates/fineract-secret.yaml:22]: Actually, it's already Kind = Secret so I think that might be a bit redundant...

**Comment by Yann-J** [comment]
OK, I've made an update that addresses most comments above (I think), in particular avoid setting default passwords, and raising an error if they're empty.

Some tuning along the way, in particular all extra variables (Secret or not) are now evaluated as templates - don't know if that's useful here but it doesn't hurt, and I've seen many cases where it is. They're also all supporting multiline values. Again, not sure if it's super useful here but you never know.

**Comment by Yann-J** [comment]
> @BLasan have you noticed this PR? It looks like something you could be interested in helping to review... 😄
> 
> I was thinking about something when reviewing #1653: That (resource limits) should be done here as well? More importantly, we're now maintaining Kubernetes YAML in 2 places (original raw Kube YAML, and Helm chart templates). That's perhaps not ideal... you guys should figure out how you want to do this going forward? Keep both? Replace the original YAML with the Helm templates, and subsequently remove the originals? Up to you - work together.

Regarding the scripts vs Helm, I would say that Helm is probably the de-facto standard nowadays for deploying to k8s, but not everyone will have it, so it could make sense to keep supporting both.

However, I would say that in the long run, once it's quite stable, the Helm chart probably should be in a separate repo so it can be versioned independently from the application, and then published into https://artifacthub.io/

File: helm/fineract/values.yaml
Comment (Yann-J) [reply helm/fineract/values.yaml]: Yeah... I would argue that the situation is a bit different in a Helm context since overriding parameters is dead simple (`helm install --set mysql.auth.password=...`), and really part of the Helm workflow, so the resistance is probably a bit lower there...

Anyway my latest commit is addressing this, and provides an error message in case the values aren't provided.

**Comment by Yann-J** [comment]
Note that there would be a lot more we could do... in particular the next item on my list is to implement a `helm test` hook, as well as making the `mysql` dependency optional, in case you already have an external instance running.

This is really only a quickstart...

**Comment by BLasan** [comment]
> Note that there would be a lot more we could do... in particular the next item on my list is to implement a `helm test` hook, as well as making the `mysql` dependency optional, in case you already have an external instance running.
> 
> This is really only a quickstart...



> > @BLasan have you noticed this PR? It looks like something you could be interested in helping to review... smile
> > I was thinking about something when reviewing #1653: That (resource limits) should be done here as well? More importantly, we're now maintaining Kubernetes YAML in 2 places (original raw Kube YAML, and Helm chart templates). That's perhaps not ideal... you guys should figure out how you want to do this going forward? Keep both? Replace the original YAML with the Helm templates, and subsequently remove the originals? Up to you - work together.
> 
> Regarding the scripts vs Helm, I would say that Helm is probably the de-facto standard nowadays for deploying to k8s, but not everyone will have it, so it could make sense to keep supporting both.
> 
> However, I would say that in the long run, once it's quite stable, the Helm chart probably should be in a separate repo so it can be versioned independently from the application, and then published into https://artifacthub.io/

+1 @vorburger I think this is ideal as we can maintain the helm chart implmentations separately and do releases accordingly

**Comment by BLasan** [comment]
> Note that there would be a lot more we could do... in particular the next item on my list is to implement a `helm test` hook, as well as making the `mysql` dependency optional, in case you already have an external instance running.
> 
> This is really only a quickstart...

Great. Shall we wait until @vorburger and @edcable  's reviews ?

File: helm/fineract/templates/fineract-secret.yaml
Comment (BLasan) [reply helm/fineract/templates/fineract-secret.yaml:22]: Yuh, but you have used the prefix `{{ .Release.Name }}` for other resources. So instead of having just the `{{ .Release.Name }}` I would suggest to use `{{ .Release.Name }}-secret` as the secret name. That'll be something meaningful. But it's not a requirement with high priority. Have you done a release for this implementation? Hope this works fine :) The configurations you've done look good to me

**Comment by Yann-J** [comment]
Is there anything else you folks need from me on this one?

**Comment by awasum** [comment]
@Yann-J  There are some Apache Rat failures in the build on Travis: https://travis-ci.com/github/apache/fineract/builds/220288308

You could either add the licenses or ignore the files. After that @vorburger may give his go ahead or more feedback...

**Pull request closed without merging**
